In [1]:
import pandas as pd
import numpy as np

In [2]:
train = pd.read_csv("./smhrd_data/Train.csv")
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6999 entries, 0 to 6998
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   6999 non-null   int64  
 1   Warehouse_block      6999 non-null   object 
 2   Mode_of_Shipment     6999 non-null   object 
 3   Customer_care_calls  5423 non-null   float64
 4   Customer_rating      6999 non-null   int64  
 5   Cost_of_the_Product  6999 non-null   int64  
 6   Prior_purchases      6049 non-null   float64
 7   Product_importance   6999 non-null   object 
 8   Gender               6999 non-null   object 
 9   Discount_offered     3468 non-null   float64
 10  Weight_in_gms        6999 non-null   object 
 11  Reached.on.Time_Y.N  6999 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 656.3+ KB


In [3]:
test = pd.read_csv("./smhrd_data/test.csv")
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   4000 non-null   int64  
 1   Warehouse_block      4000 non-null   object 
 2   Mode_of_Shipment     4000 non-null   object 
 3   Customer_care_calls  3144 non-null   float64
 4   Customer_rating      4000 non-null   int64  
 5   Cost_of_the_Product  4000 non-null   int64  
 6   Prior_purchases      3633 non-null   float64
 7   Product_importance   4000 non-null   object 
 8   Gender               4000 non-null   object 
 9   Discount_offered     2085 non-null   float64
 10  Weight_in_gms        4000 non-null   object 
dtypes: float64(3), int64(3), object(5)
memory usage: 343.9+ KB


In [ ]:
train.apply(lambda row: any)

In [14]:
mask = train.apply(lambda row: any(['?' in str(val) for val in row]), axis=1)

# Use the mask to filter the rows containing '?'
rows_with_question_mark = train[mask]

# Print the rows containing '?'
print(rows_with_question_mark)

        ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
15      16                B                ?                  5.0   
23      24                C                ?                  4.0   
32      33                F             Road                  NaN   
41      42                C           Flight                  NaN   
47      48                D             Ship                  NaN   
...    ...              ...              ...                  ...   
6945  6946                D             Ship                  2.0   
6972  6973                F             Road                  NaN   
6973  6974                C             Road                  NaN   
6975  6976                F             Ship                  NaN   
6989  6990                F             Ship                  5.0   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
15                  5                  161              3.0   
23                  2                  273   

In [15]:
df = train.replace('?', pd.NA)

# Create a boolean mask where NaN is present in any column of each row
mask = df.isna().any(axis=1)

# Use the mask to filter the rows containing NaN
rows_with_nan = df[mask]

# Print the rows containing NaN
print(rows_with_nan)

        ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
0        1                C             Ship                  2.0   
2        3                F             Ship                  3.0   
3        4                F             Ship                  3.0   
4        5                D           Flight                  NaN   
5        6                B             Ship                  4.0   
...    ...              ...              ...                  ...   
6993  6994                B             Ship                  6.0   
6995  6996                F             Ship                  3.0   
6996  6997                D             Ship                  NaN   
6997  6998                A             Road                  3.0   
6998  6999                A             Ship                  3.0   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   2                  135              3.0   
2                   1                  229   

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6999 entries, 0 to 6998
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   6999 non-null   int64  
 1   Warehouse_block      6999 non-null   object 
 2   Mode_of_Shipment     6626 non-null   object 
 3   Customer_care_calls  5423 non-null   float64
 4   Customer_rating      6999 non-null   int64  
 5   Cost_of_the_Product  6999 non-null   int64  
 6   Prior_purchases      6049 non-null   float64
 7   Product_importance   6902 non-null   object 
 8   Gender               6999 non-null   object 
 9   Discount_offered     3468 non-null   float64
 10  Weight_in_gms        6553 non-null   object 
 11  Reached.on.Time_Y.N  6999 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 656.3+ KB


In [20]:
df.describe()

ID  Customer_care_calls  Customer_rating  Cost_of_the_Product  \
count  6999.000000          5423.000000      6999.000000          6999.000000   
mean   3500.000000             4.054582         3.023718           214.249178   
std    2020.581599             1.131019         2.149861           208.294273   
min       1.000000             2.000000         1.000000            96.000000   
25%    1750.500000             3.000000         2.000000           169.000000   
50%    3500.000000             4.000000         3.000000           214.000000   
75%    5249.500000             5.000000         4.000000           251.000000   
max    6999.000000             7.000000        99.000000          9999.000000   

       Prior_purchases  Discount_offered  Reached.on.Time_Y.N  
count      6049.000000       3468.000000          6999.000000  
mean          3.576294         13.269031             0.594656  
std           1.530933         16.070188             0.490993  
min           2.000000          1.000000             0.000000  
25%           3.000000          4.000000             0.000000  
50%           3.000000          7.000000             1.000000  
75%           4.000000         10.000000             1.000000  
max          10.000000         65.000000             1.000000

In [22]:
df.head()

ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  Customer_rating  \
0   1                C             Ship                  2.0                2   
1   2                F           Flight                  5.0                3   
2   3                F             Ship                  3.0                1   
3   4                F             Ship                  3.0                2   
4   5                D           Flight                  NaN                1   

   Cost_of_the_Product  Prior_purchases Product_importance Gender  \
0                  135              3.0                low      M   
1                  225              5.0                low      F   
2                  229              3.0                low      M   
3                  228              NaN             medium      M   
4                  195              6.0               high      F   

   Discount_offered Weight_in_gms  Reached.on.Time_Y.N  
0               NaN          1339                    1  
1               6.0          1082                    0  
2               NaN          4971                    1  
3               2.0          5640                    0  
4               NaN          4944                    1

In [23]:
df.describe()

ID  Customer_care_calls  Customer_rating  Cost_of_the_Product  \
count  6999.000000          5423.000000      6999.000000          6999.000000   
mean   3500.000000             4.054582         3.023718           214.249178   
std    2020.581599             1.131019         2.149861           208.294273   
min       1.000000             2.000000         1.000000            96.000000   
25%    1750.500000             3.000000         2.000000           169.000000   
50%    3500.000000             4.000000         3.000000           214.000000   
75%    5249.500000             5.000000         4.000000           251.000000   
max    6999.000000             7.000000        99.000000          9999.000000   

       Prior_purchases  Discount_offered  Reached.on.Time_Y.N  
count      6049.000000       3468.000000          6999.000000  
mean          3.576294         13.269031             0.594656  
std           1.530933         16.070188             0.490993  
min           2.000000          1.000000             0.000000  
25%           3.000000          4.000000             0.000000  
50%           3.000000          7.000000             1.000000  
75%           4.000000         10.000000             1.000000  
max          10.000000         65.000000             1.000000

In [24]:
df[df['Customer_rating']>5]

ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
3150  3151                D             Ship                  5.0   
6394  6395                C           Flight                  4.0   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
3150               99                  255              NaN   
6394               99                  180              3.0   

     Product_importance Gender  Discount_offered Weight_in_gms  \
3150               high      M               NaN          4044   
6394                low      F               NaN          1434   

      Reached.on.Time_Y.N  
3150                    1  
6394                    1

In [33]:
df[df['Cost_of_the_Product']>350]

ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
111    112                C             Ship                  3.0   
6255  6256                B             Ship                  2.0   
6612  6613                F           Flight                  4.0   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
111                 2                 9999              4.0   
6255                2                 9999              4.0   
6612                4                 9999              3.0   

     Product_importance Gender  Discount_offered Weight_in_gms  \
111              medium      F               NaN          1322   
6255             medium      M               NaN          4066   
6612             medium      F               NaN          5555   

      Reached.on.Time_Y.N  
111                     1  
6255                    0  
6612                    1

### Customer_rating에는 이상치 99가 있다 -> Customer_care_calls 몇 이상? 인 사람들의 평균 Customer_rating로 바꾸면 되지 않을까?
### 결측치도 있지만 ?인 컬럼도 존재한다
###  Discount_offered 는 결측치가 너무 많아서 사용할 수 없을듯하다.
### Gender는 중요도가 그다지 높아보이지 않는다.
### customer_care_calls 는 도착여부와 관련이 있는듯함.  -> 전화를 많이해서 도착을 한것일까, 도착을 많이 안해서 전화를 많이 한것인가?
### prior_purchases는 겉으로는 중요해보이지 않지만 회사 내부에서 vip 고객일수록 배송을 우선시한다는 가정을 해본다면 prior_purchases구간화를 진행하고 도착여부에 따른 groupby를 진행해서 채울 수 있지 않을까?
### Mode_of_shipment는 Weight_in_gms와 warehouse_block 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6999 entries, 0 to 6998
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   6999 non-null   int64  
 1   Warehouse_block      6999 non-null   object 
 2   Mode_of_Shipment     6626 non-null   object 
 3   Customer_care_calls  5423 non-null   float64
 4   Customer_rating      6999 non-null   int64  
 5   Cost_of_the_Product  6999 non-null   int64  
 6   Prior_purchases      6049 non-null   float64
 7   Product_importance   6902 non-null   object 
 8   Gender               6999 non-null   object 
 9   Discount_offered     3468 non-null   float64
 10  Weight_in_gms        6553 non-null   object 
 11  Reached.on.Time_Y.N  6999 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 656.3+ KB

## 전처리 관련
- 1. 결측치 null 대신 ?로 표시되어 있는 데이터들을 nan으로 변경
- 2. Discount_offered는 결측치가 너무 많아 사용하기 힘들듯하다
- 3. Mode_of_Shipment, Product_importance에 가벼운 오타를 수정필요
- 4. 결측치 채우는 순서는 결측치가 적은 컬럼부터 채우는것이 맞지 않을까? (결측치 컬럼을 채울 때 결측치가 있는 컬럼을 참고할 수는 없기 때문?, 만약 결측치가 없는 컬럼을 보고 채운다면 얘기가 다름)
- 5. 위의 4번대로라면 결측치 채울 순서는 Product_importance, Mode_of_Shipment, Weight_in_gms, Prior_purchases, Customer_care_calls
- 6. 범주형 변수와 연속형 변수가 섞여있기 때문에 변수 타입이 섞여있을 때의 상관관계를 볼 수 있을지 알아봐야할듯 (Point biserial correlation라는 것이 있다하는데 scipy.stats라이브러리 알아봐야할듯)

- customer_care_calls 는 도착여부와 관련이 있는듯함.  -> 전화를 많이해서 도착을 한것일까, 도착을 많이 안해서 전화를 많이 한것인가?
- prior_purchases는 겉으로는 중요해보이지 않지만 회사 내부에서 vip 고객일수록 배송을 우선시한다는 가정을 해본다면 prior_purchases구간화를 진행하고 도착여부에 따른 groupby를 진행해서 채울 수 있지 않을까?

In [30]:
for i in [1,2,7,8]:
    print(df.iloc[:, i].value_counts())

F    2311
C    1184
A    1174
D    1168
B    1162
Name: Warehouse_block , dtype: int64
 Ship        4508
 Road        1075
 Flight      1039
 Shipzk         2
 Flightzk       1
 Roadzk         1
Name: Mode_of_Shipment, dtype: int64
low        3344
medium     2982
high        573
mediumm       1
loww          1
highh         1
Name: Product_importance, dtype: int64
M    3509
F    3490
Name: Gender, dtype: int64


In [38]:
import pandas as pd
import numpy as np

train = pd.read_csv("./smhrd_data/Train.csv")
train_df = train.replace('?', pd.NA)

# Create a boolean mask where NaN is present in any column of each row
mask = train_df.isna().any(axis=1)

# Use the mask to filter the rows containing NaN
rows_with_nan = train_df[mask]

# Print the rows containing NaN
print(rows_with_nan)

        ID Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
0        1                C             Ship                  2.0   
2        3                F             Ship                  3.0   
3        4                F             Ship                  3.0   
4        5                D           Flight                  NaN   
5        6                B             Ship                  4.0   
...    ...              ...              ...                  ...   
6993  6994                B             Ship                  6.0   
6995  6996                F             Ship                  3.0   
6996  6997                D             Ship                  NaN   
6997  6998                A             Road                  3.0   
6998  6999                A             Ship                  3.0   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   2                  135              3.0   
2                   1                  229   

In [42]:
train_df.iloc[45:50, :].info()

train_df.iloc[45:50, :]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 45 to 49
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   5 non-null      int64  
 1   Warehouse_block      5 non-null      object 
 2   Mode_of_Shipment     5 non-null      object 
 3   Customer_care_calls  3 non-null      float64
 4   Customer_rating      5 non-null      int64  
 5   Cost_of_the_Product  5 non-null      int64  
 6   Prior_purchases      4 non-null      float64
 7   Product_importance   5 non-null      object 
 8   Gender               5 non-null      object 
 9   Discount_offered     5 non-null      float64
 10  Weight_in_gms        4 non-null      object 
 11  Reached.on.Time_Y.N  5 non-null      int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 612.0+ bytes
